generator/restriction_collector.cpp

```diff
@@ -0,0 +1,210 @@
+      return false;
+    numbers.push_back(number);
+    ++iter;
+  }
+  return true;
+}
+}  // namespace
+
+namespace routing
+{
+RestrictionCollector::RestrictionCollector(string const & restrictionPath,
+                                           string const & featureId2OsmIdsPath)
+{
+  if (!ParseFeatureId2OsmIdsMapping(featureId2OsmIdsPath))
+    return;
+
+  if (!ParseRestrictions(restrictionPath))
+  {
+    m_restrictions.clear();
+    return;
+  }
+  ComposeRestrictions();
+  RemoveInvalidRestrictions();
+  LOG(LINFO, ("m_restrictions.size() =", m_restrictions.size()));
+}
+
+bool RestrictionCollector::IsValid() const
+{
+  return !m_restrictions.empty() &&
+         find_if(begin(m_restrictions), end(m_restrictions),
+                 [](Restriction const & r) { return !r.IsValid(); }) == end(m_restrictions);
+}
+
+bool RestrictionCollector::ParseFeatureId2OsmIdsMapping(string const & featureId2OsmIdsPath)
+{
+  ifstream featureId2OsmIdsStream(featureId2OsmIdsPath);
+  if (featureId2OsmIdsStream.fail())
+    return false;
+
+  while (featureId2OsmIdsStream)
+  {
+    string line;
+    if (!getline(featureId2OsmIdsStream, line))
+      return true;
+
+    vector<uint64_t> osmIds;
+    strings::SimpleTokenizer iter(line, ",");
+    if (!ParseLineOfNumbers(iter, osmIds))
+      return false;
+
+    if (osmIds.size() < 2)
+      return false;  // Every line should contain at least feature id and osm id.
+
+    uint32_t const featureId = static_cast<uint32_t>(osmIds.front());
+    osmIds.erase(osmIds.begin());
+    AddFeatureId(featureId, osmIds);
+  }
+  return true;
+}
+
+bool RestrictionCollector::ParseRestrictions(string const & restrictionPath)
+{
+  ifstream restrictionsStream(restrictionPath);
+  if (restrictionsStream.fail())
+    return false;
+
+  while (restrictionsStream)
+  {
+    string line;
+    if (!getline(restrictionsStream, line))
+      return true;
+
+    strings::SimpleTokenizer iter(line, ",");
+    Restriction::Type type;
+    if (!FromString(*iter, type))
+      return false;
+
+    ++iter;
+    vector<uint64_t> osmIds;
+    if (!ParseLineOfNumbers(iter, osmIds))
+      return false;
+
+    AddRestriction(type, osmIds);
+  }
+  return true;
+}
+
+void RestrictionCollector::ComposeRestrictions()
+{
+  // Going through all osm ids saved in |m_restrictionIndex| (mentioned in restrictions).
+  size_t const numRestrictions = m_restrictions.size();
+  for (auto const & osmIdAndIndex : m_restrictionIndex)
+  {
+    LinkIndex const & index = osmIdAndIndex.second;
+    CHECK_LESS(index.m_restrictionNumber, numRestrictions, ());
+    Restriction & restriction = m_restrictions[index.m_restrictionNumber];
+    CHECK_LESS(index.m_linkNumber, restriction.m_links.size(), ());
+
+    uint64_t const & osmId = osmIdAndIndex.first;
+    // Checking if there's an osm id belongs to a restriction is saved only once as feature id.
+    auto const rangeId = m_osmIds2FeatureId.equal_range(osmId);
+    if (rangeId.first == rangeId.second)
```

Comment:
This place is rewritten. Now I use map instead of unordered_multimap. And check if any duplicates on the stage of filling |m_osmId2FeatureId|.